crates/net/network/src/transactions.rs

```diff
@@ -1016,6 +1011,56 @@ struct Peer {
     client_version: Arc<String>,
 }
 
+// === TransactionFetcher ===
+/// * active requests - especially inflight hashes, so that whenever we receive new missing hashes
+///   we can check first if they're already being requested.
+/// * missing hashes and peers that send us these - so we can possibly re-request them
+
+#[derive(Debug, Default)]
+struct TransactionFetcher {
+    /// All currently active requests for pooled transactions.
+    inflight_requests: FuturesUnordered<GetPooledTxRequestFut>,
+    /// Set that tracks all hashes that are currently being fetched.
+    inflight_hash_to_fallback_peers: HashMap<TxHash, Vec<PeerId>>,
+}
+
+impl TransactionFetcher {
+    // request the missing transactions
+    fn request_from(&mut self, hashes: Vec<TxHash>, peer: &Peer) -> u64 {
+        let peer_id: PeerId = peer.request_tx.peer_id;
+        let mut missing_hashes: Vec<TxHash> = Vec::new();
+        // 1. filter out inflight hashes, and register the peer as fallback for all inflight hashes
+        for hash in hashes {
+            match self.inflight_hash_to_fallback_peers.entry(hash) {
+                Entry::Vacant(entry) => {
+                    // the hash is not in inflight hashes, insert it
+                    missing_hashes.push(*entry.key());
+                    entry.insert(vec![peer_id]);
+                }
+                Entry::Occupied(mut entry) => {
+                    // the hash is already in inflight, add this peer as a backup
+                    entry.get_mut().push(peer_id);
+                }
+            }
+        }
+        // 2. request all missing from peer
```

Comment:
here it could be that the hashes vec is empty: nothing to request and we can return early